app/controllers/orders_controller.rb

```diff
@@ -1,34 +1,138 @@
+      flash[:confirmed_order_id] = @order.id
+      # how to persist page if refreshed?
+      session[:order_id] = nil
 
-  # def create
-  #   Order.create
-  # end
+      redirect_to confirmation_path
+    else
+      flash.now[:errors] = "Your order could not be completed. See below for errors."
+
+      render :checkout
+    end
+  end
+
+  def confirmation
+    flash.keep
+    @order = Order.find(flash[:confirmed_order_id])
+    # use the below line for debugging (to avoid losing the flash[:confirmed_order_id])
+    # @order = Order.find(12)
+  end
 
-  # def purchase
-  #   order = Order.find(session[:order_id])
-  #   order.update(order_params)
-  #   if order.save
-  #     redirect_to finalize_order_path
-  #   else
-  #     render purchase
-  #   end
-  # end
+  def cancel
+    @order = Order.find(flash[:confirmed_order_id])
+    @order.status = "cancelled"
+    @order.save
+
+    redirect_to root_path
+  end
+
+  def fulfillment
+    @user = User.find(params[:id])
+    @order_items = @user.order_items
+    @total_revenue = @order_items.sum(:item_total)
+    @filtered_order_items = nil
+  end
+
+  def mark_shipped
+    @user = User.find(params[:id])
+    @order = Order.find(params[:order_id])
+    @order.mark_shipped!
+
+    redirect_to order_fulfillment_path
+  end
+
+  def merchant
+    @products_array = []
+    @merchants = User.all
+    if params[:user].empty?
+      redirect_to products_path
+    else
+      @merchant = User.find(params[:user])
+      @merchant_products = @merchant.products
+
+      render :index
+    end
+  end
+
+  def filter_status
+    if params[:status].empty?
+      redirect_to order_fulfillment_path
+    else
+      # setup for fulfillment action
+      # page still needs @order_items != nil
+      # even though not rendering it
+      @user = User.find(params[:id])
+      @order_items = @user.order_items
+      # orders displayed after filter
+      @filtered_order_items = OrderItem.joins(:order, :product).where(orders: { status: params[:status] }, products: { user_id: @user.id })
+
+      render :fulfillment
+    end
+  end
 
   private
 
-  # do I need this here? will need it for :update, not for :create
+  # ensures merchant signed in can only see their own user pages
+  def correct_merchant
+    if request.path.include?(session[:user_id].to_s) == false
+      redirect_to user_path(@user.id)
+    end
+  end
+
+  def belongs_to_user
+    @order = Order.find(params[:id])
+    @order.products.each do |x|
+      if x.user_id == @current_user.id
+        return
+      end
+    end
+
+    redirect_to order_fulfillment_path(@user.id)
+  end
+
   def order_params
-    params.require(:order).permit(:status, :email, :cc_name, :cc_number, 
+    params.require(:order).permit(:status, :email, :cc_name, :cc_number,
       :cc_expiration, :cc_cvv, :billing_zip, :shipped, :address1, :address2,
-      :city, :state, :mailing_zip)
+      :city, :state, :mailing_zip, :mailing_name)
+  end
+
+  def total_sales
```

Comment:
I like that you isolated this logic into its own method. Makes it reusable and much easier to test. However, I would consider changing how it's called. Right now, it depends on `@order` and `@user_items` to already exist and be in scope. This may not be apparent to someone new to the project (or you, six months from now). You could make this dependency obvious by requiring them as arguments, like...

```ruby
# in a another controller action
order = Order.find(params[:id])
user_items = current_user.order_items
@total_sales = total_sales_for(order, user_items)

# then the actual method def would look like
def total_sales_for(order, user_items)
  total_sales = 0
  # do the work
  return total_sales
end
```